masu/processor/providers/aws/report_downloader.py

```diff
@@ -0,0 +1,209 @@
+
+        manifest_json = None
+        with open(manifest_file, 'r') as f:
+            manifest_json = json.load(f)
+
+        return manifest_json
+
+    def _get_report_path(self, dt):
+        """
+        Return path of report files.
+
+        Args:
+            dt (DateTime): The starting datetime object
+
+        Returns:
+            (String): "/prefix/report_name/YYYYMMDD-YYYYMMDD",
+                    example: "/my-prefix/my-report/19701101-19701201"
+
+        """
+        report_date_range = self._get_manifest_date_range(dt)
+        return '{}/{}/{}'.format(self.bucket_prefix,
+                                 self.report_name,
+                                 report_date_range)
+
+    def _get_manifest_date_range(self, dt):
+        """
+        Get a formatted date range string.
+
+        Args:
+            dt (DateTime): The starting datetime object
+
+        Returns:
+            (String): "YYYYMMDD-YYYYMMDD", example: "19701101-19701201"
+
+        """
+        start_month = dt.replace(day=1, second=1, microsecond=1)
+        end_month = start_month + relativedelta(months=+1)
+        timeformat = '%Y%m%d'
+        return '{}-{}'.format(start_month.strftime(timeformat),
+                              end_month.strftime(timeformat))
+
+    def download_bucket(self):
+        """
+        Bulk Download all files in an s3 bucket.
+
+        Returns:
+            (List) List of filenames downloaded.
+
+        """
+        s3_resource = boto3.resource('s3')
+        bucket = s3_resource.Bucket(self.bucket)
+        files = []
+        for s3obj in bucket.objects.all():
+            files.append(self.download_file(s3obj.key))
+        return files
+
+    def download_current_report(self):
+        """
+        Read CUR manifest, download current report files.
+
+        Returns:
+            (String): The path and file name of the saved file
+
+        """
+        return self.download_report(datetime.today())
+
+    def download_file(self, key):
+        """
+        Download an S3 object to file.
+
+        Args:
+            key (str): The S3 object key identified.
+
+        Returns:
+            (String): The path and file name of the saved file
+
+        """
+        s3_filename = key.split('/')[-1]
+        file_path = f'{DATA_DIR}/{self.customer_schema}/aws'
+        file_name = f'{file_path}/{s3_filename}'
+        # Make sure the data directory exists
+        os.makedirs(file_path, exist_ok=True)
+
+        etag = self.report_stats.get_etag(key)
+        s3_file = self.s3.get_object(Bucket=self.bucket, Key=key)
+        if s3_file.get('etag') != etag:
+            self.logger.info('Downloading %s to %s', s3_filename, file_name)
+            self.s3.download_file(self.bucket, key, file_name)
+            self.report_stats.update(report_name=key,
+                                     etag=s3_file.get('etag'),
+                                     last_completed_datetime=datetime.now())
```

Comment:
I might be wrong in the intention, but I was thinking that the last_completed_datetime would be reserved for when the report processing was completed, not just the etag updated. I would read that as the report with that etag has already been processed.

Reply:
Okay, that's fine. 

I wanted a way to signal that we've actually downloaded new files versus just returned the list of already-cached files. I can have the Downloader return a more complex object that includes a `{changed: True}` dict. Or maybe there's a third option? Thoughts?